test/net/sourceforge/kolmafia/request/AscensionRequestTest.java

```diff
@@ -0,0 +1,151 @@
+
+      try (gashCleanups) {
+        // This does a 302 redirect to afterlife.php
+        var jumpGash = new GenericRequest("ascend.php?action=ascend&pwd&confirm=on&confirm2=on");
+        jumpGash.run();
+      }
+
+      var afterlifeCleanups = withNextResponse(200, html("request/test_ascension_jump_gash.html"));
+
+      try (afterlifeCleanups) {
+        // Execute the result of the 302 redirect and get the output which puts us in valhalla
+        var afterlife = new GenericRequest("afterlife.php");
+        afterlife.run();
+      }
+
+      // Confirm our last run breakfast is now -1
+      assertThat(breakfastPref, isSetTo(breakfastValhalla));
+      // Confirm our ascensions today has incremented
+      assertThat(ascensionsPref, isSetTo(ascensionAfter));
+    }
+  }
+
+  @Test
+  void testKarmaUpdates() {
+    // Test that our karma is updated correctly when we enter the pearly gates
+    String pref = "bankedKarma";
+    int karmaBeforehand = 100;
+    int karmaAfter = 211;
+
+    // Set our karma beforehand to 100
+    var cleanups = withProperty(pref, karmaBeforehand);
+
+    try (cleanups) {
+      // Confirm our karma is 100
+      assertThat(pref, isSetTo(karmaBeforehand));
+
+      // Execute a request to enter the pearly gates
+      var pearlyGates = new GenericRequest("afterlife.php?action=pearlygates");
+      pearlyGates.responseText = html("request/test_ascension_enter_valhalla.html");
+      pearlyGates.setHasResult(true);
+      pearlyGates.processResponse();
+
+      // Confirm our karma is now 211
+      assertThat(pref, isSetTo(karmaAfter));
+    }
+  }
+
+  @Test
+  void testSelectPathUnconfirmedUnprocessed() {
+    // Set last breakfast to -1 to mark that we've ascended as done in testAscensionsTodayTracked
+    var cleanups =
+        new Cleanups(withProperty("lastBreakfast", -1), withProperty("bankedKarma", 150));
+    ByteArrayOutputStream ostream = new ByteArrayOutputStream();
+
+    try (cleanups;
+        PrintStream out = new PrintStream(ostream, true)) {
+      RequestLogger.openCustom(out);
+
+      // Testing to ensure "Are you sure" is not handled
+      var ascendRequest =
+          new GenericRequest(
+              "afterlife.php?action=ascend&asctype=2&whichclass=1&gender=2&whichpath=44&whichsign=3");
+      ascendRequest.setHasResult(true);
+      ascendRequest.run();
+
+      RequestLogger.closeCustom();
+    }
+
+    // Confirm unchanged, we did not actually ascend yet
+    String logged = ostream.toString().trim();
+    assertEquals("", logged);
+  }
+
+  @Test
+  void testSelectPathConfirmedProcessed() {
+    // Set last breakfast to -1 to mark that we've ascended as done in testAscensionsTodayTracked
+    var cleanups =
+        new Cleanups(
+            withProperty("lastBreakfast", -1),
+            withProperty("bankedKarma", 150),
+            withNextResponse(302, Map.of("location", List.of("main.php")), ""));
```

Comment:
Need to create another test where ascending does not place you into choice.php, and fix this test to place you into choice.php

Reply:
👍

Reply:
Done, although I'm not too happy with the way I went about things.
I'm reusing Grey You, but although it should be fine; The path ultimately does resolve into a choice which could in theory, play up at a later date.